x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SparseScoredTermsMapper.java

```diff
@@ -0,0 +1,288 @@
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.index.analysis.NamedAnalyzer;
+import org.elasticsearch.index.mapper.DocumentParserContext;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.MapperBuilderContext;
+import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.SourceValueFetcher;
+import org.elasticsearch.index.mapper.TermBasedFieldType;
+import org.elasticsearch.index.mapper.TextSearchInfo;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentSubParser;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+public class SparseScoredTermsMapper extends FieldMapper {
+    public static final String CONTENT_TYPE = "sparse_scored_terms";
+    private final Explicit<Boolean> ignoreMalformed;
+    private final boolean ignoreMalformedByDefault;
+
+    public static final TypeParser PARSER = new TypeParser(
+        (n, c) -> new SparseScoredTermsMapper.Builder(n, IGNORE_MALFORMED_SETTING.get(c.getSettings()))
+    );
+
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(true);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS_AND_FREQS);
+            FIELD_TYPE.freeze();
+        }
+        public static TextSearchInfo TEXT_SEARCH_INFO = new TextSearchInfo(
+            FIELD_TYPE,
+            null,
+            Lucene.KEYWORD_ANALYZER,
+            Lucene.KEYWORD_ANALYZER
+        );
+    }
+
+    public SparseScoredTermsMapper(
+        String simpleName,
+        MappedFieldType mappedFieldType,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        Builder builder
+    ) {
+        super(simpleName, mappedFieldType, multiFields, copyTo);
+        this.ignoreMalformed = builder.ignoreMalformed.getValue();
+        this.ignoreMalformedByDefault = builder.ignoreMalformed.getDefaultValue().value();
+    }
+
+    private record SparseTermsAndScores(List<String> terms, List<Integer> scores) {
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<SparseTermsAndScores, Void> SPARSE_TERMS_MAPPED_FIELD_PARSER =
+        new ConstructingObjectParser<>(
+            "sparse_terms_mapped_field",
+            false,
+            a -> new SparseTermsAndScores((List<String>) a[0], (List<Integer>) a[1])
+        );
+    static {
```

Comment:
Another option for the format is 
```
[
{"term": "foo", "score": 2},
{"term": "bar", "score": 100},
...
]
```
We would probably need a top level object field name so that we can create the token stream with the term frequency.